google-cloud-firestore/lib/google/cloud/firestore/field_path.rb

```diff
@@ -0,0 +1,211 @@
+        end
+
+        ##
+        # Creates a field path object representing the sentinel ID of a
+        # document. It can be used in queries to sort or filter by the document
+        # ID. See {Client#document_id}.
+        #
+        # @return [FieldPath] The field path object.
+        #
+        # @example
+        #   require "google/cloud/firestore"
+        #
+        #   firestore = Google::Cloud::Firestore.new
+        #
+        #   # Get a collection reference
+        #   cities_col = firestore.col "cities"
+        #
+        #   # Create a query
+        #   query = cities_col.start_at("NYC").order(
+        #     Google::Cloud::Firestore::FieldPath.document_id
+        #   )
+        #
+        #   query.get do |city|
+        #     puts "#{city.document_id} has #{city[:population]} residents."
+        #   end
+        #
+        def self.document_id
+          new :__name__
+        end
+
+        ##
+        # Creates a field path object representing the nested fields for
+        # document data.
+        #
+        # @param [String] dotted_string A string representing the path of the
+        #   document data. The string can represent as a string of individual
+        #   fields joined by ".". Fields containing `~`, `*`, `/`, `[`, `]`, and
+        #   `.` cannot be in a dotted string, and should be created by passing
+        #   individual field strings to  {FieldPath.new} instead.
+        #
+        # @return [FieldPath] The field path object.
+        #
+        # @example
+        #   require "google/cloud/firestore"
+        #
+        #   firestore = Google::Cloud::Firestore.new
+        #
+        #   user_snap = firestore.doc("users/frank").get
+        #
+        #   field_path = Google::Cloud::Firestore::FieldPath.parse(
+        #     "favorites.food"
+        #   )
+        #   field_path.fields #=> ["favorites", "food"]
+        #
+        def self.parse dotted_string
+          return new dotted_string if dotted_string.is_a? Array
+
+          fields = String(dotted_string).split(".")
+          fields.each do |field|
+            if INVALID_FIELD_PATH_CHARS.match field
+              fail ArgumentError, "invalid character, use FieldPath instead"
+            end
+          end
+          new fields
+        end
+
+        ##
+        # @private
+        def <=> other
+          return nil unless other.is_a? FieldPath
+          formatted_string <=> other.formatted_string
+        end
+
+        ##
+        # @private
+        def eql? other
+          formatted_string.eql? other.formatted_string
+        end
+
+        ##
+        # @private
+        def hash
+          formatted_string.hash
+        end
+
+        protected
+
+        START_FIELD_PATH_CHARS = /\A[a-zA-Z_]/
+        INVALID_FIELD_PATH_CHARS = %r{[\~\*\/\[\]]}
+
+        def escape_field_for_path field
+          field = String field
+
+          return "`#{field}`" if INVALID_FIELD_PATH_CHARS.match field
```

Comment:
You need to escape the backtick.

Reply:
While Ruby can escape the backpack, it doesn't need to be escaped.

```ruby
"`foo`" == "\`foo\`" #=> true
```

Reply:
We need to ensure the following escaping:

```
"foo`bar" => "`foo\`bar`
"foo\bar" => `foo\\bar`
"foo\`bar" => `foo\\\`bar`
```

Reply:
Precisely which characters in the individual field node need to be escaped?

Reply:
Only the ones shown here - backslash and backtick

Reply:
Is this what you are looking for?

```ruby
def escape_field_for_path field
  field = String field

  if INVALID_FIELD_PATH_CHARS.match(field) ||
     field["."] || field["`"] || field["\\"]
    escaped_field = field.gsub /[\`\\]/, "`" => "\\\`", "\\" => "\\\\"
    return "`#{escaped_field}`"
  end

  return field if START_FIELD_PATH_CHARS.match field

  "`#{field}`"
end
```

It is only escaping <code>&#96;</code> and `\`

Reply:
That looks right to me.